Allow setting None as Initial widget value for st.selectbox, st.date_input, st.time_input, st.number_input, st.text_input, st.text_area, st.radio

sfc-gh-tszerszen opened the pull request:
## 📚 Context

_Please describe the project or issue background here_

- What kind of change does this PR introduce?

  - [x] Feature

## 🧠 Description of Changes

- _Add bullet points summarizing your changes here_

This PR allows setting `None` as initial widget value for the following widgets:
* `st.selectbox`
* `st.date_input`
* `st.time_input`
* `st.number_input`
* `st.text_input`
* `st.text_area`
* `st.radio`

Although branch name is **allow-none-as-initial-value-for-text-widgets**, it actually allows setting None on all widgets listed above, not only text widgets.

For example,  now `st.text_input` and `st.text_area` will always return "" if it’s empty, no matter if the user never wrote something or just deleted their text. This PR changes this behaviour by allowing using `None` as a value for input widgets and adds an option that allows users to clear the value in input text widgets to `None`. 

**Revised:**

![image](https://user-images.githubusercontent.com/78742618/235106362-e86a8c4f-654d-40a7-b6b8-3c38253f6e20.png)
![image](https://user-images.githubusercontent.com/78742618/235106475-c255d8a8-90a3-4736-b56b-25e31893f76d.png)


_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [x] Screenshots included


- Closes: https://github.com/streamlit/streamlit/issues/6719
- Closes: https://github.com/streamlit/streamlit/issues/5220
- Closes: https://github.com/streamlit/streamlit/issues/2325
- Closes: https://github.com/streamlit/streamlit/issues/949
- Closes: https://github.com/streamlit/streamlit/issues/4650
- Closes: https://github.com/streamlit/streamlit/issues/6742

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch commented on lib/streamlit/elements/text_widgets.py:
```diff
@@ -84,8 +85,9 @@ def text_input(
+        clearable: bool = False,
         label_visibility: LabelVisibility = "visible",
-    ) -> str:
+    ) -> Union[str, None]:
```
We probably need some overloads here. Otherwise, this will have a bit of a negative impact on the typing behavior. E.g. only if initial value is `None`, it returns an `Union[str, None]` otherwise it returns `str`.

sfc-gh-tszerszen replied:
Hey, just tried to the the following overloads:

```python
    @overload
    def text_input(
        self,
        label: str,
        value: SupportsStr,
        max_chars: Optional[int],
        key: Optional[Key],
        type: Literal["default", "password"],
        help: Optional[str],
        autocomplete: Optional[str],
        on_change: Optional[WidgetCallback],
        args: Optional[WidgetArgs],
        kwargs: Optional[WidgetKwargs],
        *,  # keyword-only arguments:
        placeholder: Optional[str],
        disabled: bool,
        clearable: bool,
        label_visibility: LabelVisibility,
    ) -> str:
        pass

    @overload
    def text_input(
        self,
        label: str,
        value: None,
        max_chars: Optional[int],
        key: Optional[Key],
        type: Literal["default", "password"],
        help: Optional[str],
        autocomplete: Optional[str],
        on_change: Optional[WidgetCallback],
        args: Optional[WidgetArgs],
        kwargs: Optional[WidgetKwargs],
        *,  # keyword-only arguments:
        placeholder: Optional[str],
        disabled: bool,
        clearable: bool,
        label_visibility: LabelVisibility,
    ) -> Union[str, None]:
        pass
```

I've tried multiple variations of the above, but `mypy` keeps failing with the error:
```bash
 No overload variant matches argument type "str" 
Possible overload variants:
e2e/scripts/app_hotkeys.py:17:1: note:     def text_input(self, label: str, value: SupportsStr, max_chars: Optional[int], key: Union[str, int, None], type: Literal['default', 'password'], help: Optional[str], autocomplete: Optional[str], on_change: Optional[Callable[..., None]], args: Optional[Tuple[Any, ...]], kwargs: Optional[Dict[str, Any]], *, placeholder: Optional[str], disabled: bool, clearable: bool, label_visibility: Literal['visible', 'hidden', 'collapsed']) -> str
e2e/scripts/app_hotkeys.py:17:1: note:     def text_input(self, label: str, value: None, max_chars: Optional[int], key: Union[str, int, None], type: Literal['default', 'password'], help: Optional[str], autocomplete: Optional[str], on_change: Optional[Callable[..., None]], args: Optional[Tuple[Any, ...]], kwargs: Optional[Dict[str, Any]], *, placeholder: Optional[str], disabled: bool, clearable: bool, label_visibility: Literal['visible', 'hidden', 'collapsed']) -> Optional[str]

```

I am not sure how to handle overloads in this case, seems like it is no enough difference for `mypy` to recognize this

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -147,6 +147,10 @@ def is_array_value_field_name(obj: object) -> TypeGuard[ArrayValueFieldName]:
     return obj in ARRAY_VALUE_FIELD_NAMES
 
 
+def check_type(val: Any) -> type:
```
I think this function makes the internal typing a bit worse. Just using `type(val) is str` is giving mypy the information that `val` has to be string. I think with `check_type`, this is not working (however, I haven't tested it).

I would just directly use `isinstance(s, str)` above to check the type.

sfc-gh-tszerszen replied:
@LukasMasuch true, should have use `isinstance` right away! Removed the func and replaced `check_type` with `isintance` check, thank you 🙂

lukasmasuch commented on frontend/src/components/widgets/TextInput/TextInput.tsx:
```diff
@@ -200,6 +200,7 @@ class TextInput extends React.PureComponent<Props, State> {
         <UIInput
           value={value}
           placeholder={placeholder}
+          clearable={element.clearable}
```
nit: I think we can also set `clearOnEscape={element.clearable}` here as well.

sfc-gh-tszerszen replied:
Good call out, it's updated, thank you 👍

lukasmasuch commented on frontend/src/components/widgets/TextArea/TextArea.tsx:
```diff
@@ -208,6 +208,7 @@ class TextArea extends React.PureComponent<Props, State> {
           <UITextArea
             data-testid="stTextArea"
             value={value}
+            clearable={element.clearable}
```
nit: I think we can also set `clearOnEscape={element.clearable}` here as well.

sfc-gh-tszerszen replied:
True, it's updated, same as above, thank you 👍

lukasmasuch commented on lib/streamlit/elements/text_widgets.py:
```diff
@@ -269,13 +278,19 @@ def _text_input(
 
         self.dg._enqueue("text_input", text_input_proto)
+        if value is None or (
+            check_type(widget_state.value) is str and widget_state.value == "None"
```
Why is the `widget_state.value` `None` here? I would assume the value is just an empty string or?

sfc-gh-tszerszen replied:
Sometimes it will be an empty `string`, but what if the value is `None`? It should follow this conditional then

sfc-gh-tszerszen commented:
@LukasMasuch thank you for your suggestions, just introduced them, I am most confused about `mypy` and `overload`. I was unable to provide a working example of such and IDK if that is even possible

sfc-gh-tszerszen commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -171,14 +171,24 @@
+        self, ui_value: Optional[str], widget_id: Any = ""
+    ) -> Union[time, None]:
+        # if not ui_value:
+        #    return None
```
Fixed in the latest commit 👍

sfc-gh-tszerszen commented on lib/streamlit/elements/number_input.py:
```diff
@@ -245,11 +249,8 @@
-            )
-        else:
+        # if value is not None:
+        if True:
```
it's fixed

sfc-gh-tszerszen commented on frontend/src/lib/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -206,7 +212,7 @@
     const { width, element, disabled, theme, widgetMgr } = this.props
     const { values, isRange } = this.state
-    const { colors, fontSizes, lineHeights } = theme
+    const { colors, fontSizes, lineHeights, spacing } = theme
```
resolved 👍

lukasmasuch commented on lib/streamlit/elements/radio.py:
```diff
@@ -77,7 +89,7 @@ def radio(
         label: str,
         options: OptionSequence[T],
-        index: int = 0,
+        index: Optional[Union[int, None]] = 0,
```
`Optional[Union[int, None]]` is a bit redundant since `Optional[int]` is the same as `Union[int, None]`. I would suggest to directly use the newer annotation form: `int | None` here instead. But you need to first import this: `from __future__ import annotations` on the top of the file.

sfc-gh-tszerszen replied:
Thank you @LukasMasuch , I've followed your recommendations for all the widgets and replaced old notation with a new one ex. `int | None` 👍 the PR is updated

lukasmasuch commented on lib/streamlit/elements/radio.py:
```diff
@@ -201,7 +213,7 @@ def _radio(
         label: str,
         options: OptionSequence[T],
-        index: int = 0,
+        index: Optional[Union[int, None]] = 0,
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `int | None`

lukasmasuch commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -73,7 +83,7 @@ def selectbox(
         label: str,
         options: OptionSequence[T],
-        index: int = 0,
+        index: Optional[Union[int, None]] = 0,
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `int | None`

lukasmasuch commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -184,7 +194,7 @@ def _selectbox(
         label: str,
         options: OptionSequence[T],
-        index: int = 0,
+        index: Optional[Union[int, None]] = 0,
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `int | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -44,16 +54,22 @@
 
-def _parse_date_value(value: DateValue) -> Tuple[List[date], bool]:
+def _parse_date_value(
+    value: Optional[Union[DateValue, str, None]]
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `DateValue | str | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -121,10 +137,10 @@ def _parse_max_date(
 @dataclass(frozen=True)
 class _DateInputValues:
-    value: Sequence[date]
+    value: Optional[Union[Sequence[date], None]]
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `Sequence[date] | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -121,10 +137,10 @@ def _parse_max_date(
     is_range: bool
-    max: date
-    min: date
+    max: Optional[Union[date, None]]
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `date | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -121,10 +137,10 @@ def _parse_max_date(
-    max: date
-    min: date
+    max: Optional[Union[date, None]]
+    min: Optional[Union[date, None]]
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `date | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -399,7 +425,7 @@ def _time_input(
         self,
         label: str,
-        value: DateValue = None,
+        value: Optional[Union[DateValue, str, None]] = "...",
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `DateValue | str | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -512,7 +538,7 @@ def date_input(
         self,
         label: str,
-        value: DateValue = None,
+        value: Optional[Union[DateValue, str, None]] = "...",
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `DateValue | str | None`

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -191,20 +207,25 @@ def deserialize(
         widget_id: str = "",
-    ) -> DateWidgetReturn:
-        return_value: Sequence[date]
+    ) -> Optional[Union[DateWidgetReturn, None]]:
```
See [comment](https://github.com/streamlit/streamlit/pull/6596/files#r1193625624), replace with: `DateWidgetReturn | None`

lukasmasuch commented on frontend/src/lib/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -328,6 +334,27 @@ class DateInput extends React.PureComponent<Props, State> {
+                            // Since the close icon is an SVG, and we can't control its viewbox nor its attributes,
+                            // Let's use a scale transform effect to make it bigger.
+                            // The width property only enlarges its bounding box, so it's easier to click.
+                            transform: "scale(1.41)",
```
What is the reason why we can't use the same scale here as used in mutliselect (`1.5`)? Does it result in different sizes?

sfc-gh-tszerszen replied:
Yes, it results in different sizes 👍 thus I had to use different values here

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -200,7 +213,25 @@ class Selectbox extends React.PureComponent<Props, State> {
+                      // Since the close icon is an SVG, and we can't control its viewbox nor its attributes,
+                      // Let's use a scale transform effect to make it bigger.
+                      // The width property only enlarges its bounding box, so it's easier to click.
+                      transform: "scale(1.25)",
```
Similar to the other comment, why do we need different scales here? Inconsistencies in baseweb?

sfc-gh-tszerszen replied:
it results in different sizes 👍 thus I had to use different values here

lukasmasuch commented on frontend/src/lib/components/widgets/Multiselect/Multiselect.tsx:
```diff
@@ -37,6 +36,7 @@ import { Placement } from "src/lib/components/shared/Tooltip"
 import { VirtualDropdown } from "src/lib/components/shared/Dropdown"
 import { fuzzyFilterSelectOptions } from "src/lib/components/shared/Dropdown/Selectbox"
 import { labelVisibilityProtoValueToEnum } from "src/lib/util/utils"
+import { withTheme } from "@emotion/react"
```
nit: was this import move intended?

sfc-gh-tszerszen replied:
Yes, I wanted it to be next to `EmotionTheme`. Do we need to introduce some import ordering tool?

lukasmasuch replied:
Hmm, I see. Yep, I think an import ordering tool would be great. The thing is that it is more consistent with the current codebase if it is on top since it's mostly sorted based on external imports first, and then internal, and last are the relative imports.

lukasmasuch commented on frontend/src/lib/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -121,11 +128,17 @@ class NumberInput extends React.PureComponent<Props, State> {
-      this.commitWidgetValue({ fromUi: false })
-    })
+    const value = this.props.element.value
+    if (value === null || value === undefined) {
```
nit: we have `notNullOrUndefined` and `isNullOrUndefined` in the utils.ts [here](https://github.com/streamlit/streamlit/blob/a177e6604c25edaac2d6570d8127bc7dfb0c1914/frontend/src/lib/util/utils.ts#L240). This could make those checks slightly more convenient.

lukasmasuch commented on frontend/src/lib/components/widgets/Selectbox/Selectbox.tsx:
```diff
@@ -103,14 +114,22 @@ class Selectbox extends React.PureComponent<Props, State> {
 
-  private onChange = (value: number): void => {
-    this.setState({ value }, () => this.commitWidgetValue({ fromUi: true }))
+  private onChange = (value?: number | undefined): void => {
```
nit: I think using `?` here already implies that it can be `undefined` so `| undefined` can be removed.

lukasmasuch commented on frontend/src/lib/components/widgets/Selectbox/Selectbox.tsx:
```diff
@@ -125,15 +144,17 @@ class Selectbox extends React.PureComponent<Props, State> {
         width={this.props.width}
         onChange={this.onChange}
-        value={this.state.value}
+        value={this.state.value !== -1 ? this.state.value : undefined}
```
Why do we need to use `-1` here (for selectbox) and not just `undefined`?

sfc-gh-tszerszen replied:
`-1` means no option is Selected, I did not make `Selectbox` index field optional, like I did with Number input, to limit proto changes

sfc-gh-mnowotka replied:
nit: value={this.state.value !== -1 || undefined}

lukasmasuch commented on frontend/src/lib/components/widgets/Selectbox/Selectbox.tsx:
```diff
@@ -103,14 +114,22 @@ class Selectbox extends React.PureComponent<Props, State> {
-  private onChange = (value: number): void => {
-    this.setState({ value }, () => this.commitWidgetValue({ fromUi: true }))
+  private onChange = (value?: number | undefined): void => {
+    if (value === undefined) {
```
Is this `if` actually needed, since `commitWidgetValue` seems also to handle this case of undefined value as well?

lukasmasuch commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -204,19 +206,20 @@ def _selectbox(
 
-        if not isinstance(index, int):
+        clearable = index is None
+        if not isinstance(index, int) and not clearable:
```
nit: I think `if not isinstance(index, int) and index is not None` would be slightly more future proof in case we add clearable as a parameter in the future.

lukasmasuch commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -204,19 +206,20 @@ def _selectbox(
         selectbox_proto = SelectboxProto()
         selectbox_proto.label = label
-        selectbox_proto.default = index
+        selectbox_proto.default = -1 if index is None else index
```
Why do we need to use `-1` here, wouldn't it be possible to just use `None` as value or just not set it at all?

sfc-gh-tszerszen replied:
`-1` means no option is Selected, I did not make Selectbox index field optional, like I did with Number input, to limit proto changes

lukasmasuch replied:
Hmm, I think we could consider using `optional` for the values in all widget proto messages for consistency. I think this change would be fine to do on all our proto messages. Using some magic values to indicate a null state feels a bit hacky. It's fine if we don't have a better workaround, but in this case `optional` seems to be better here.

lukasmasuch commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -50,9 +51,11 @@
 class SelectboxSerde(Generic[T]):
     options: Sequence[T]
-    index: int
+    index: Optional[int]
```
nit: this is preferred with future annotations: `int | None`

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -135,20 +141,26 @@ class Selectbox extends React.PureComponent<Props, State> {
+    const { label, labelVisibility, help, clearable, theme } = this.props
     let { disabled, options } = this.props
 
+    const noOptionsMsg = "No options to select."
```
nit: I think these two constants can be moved to the top of the file

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -182,13 +194,14 @@ class Selectbox extends React.PureComponent<Props, State> {
           onInputChange={this.onInputChange}
           onClose={this.onClose}
+          clearable={clearable}
+          placeholder={clearable ? chooseOptionMsg : null}
```
nit: I don't think the placeholder needs to be set conditionally or? We can probably always set the same placeholder, but it most likely will only be visible to the user in the current case of `clearable`. Later on, we might allow having a selectbox be empty as initial value, but not clearable. In that case, we still would need a placeholder.

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -182,13 +194,14 @@ class Selectbox extends React.PureComponent<Props, State> {
+          onChange={this.onChange as (params: OnChangeParams) => unknown}
           onInputChange={this.onInputChange}
           onClose={this.onClose}
+          clearable={clearable}
```
nit: we probably can also set `escapeClearsValue={clearable}` here as well

sfc-gh-tszerszen replied:
Updated 👍

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -95,6 +99,8 @@ class Selectbox extends React.PureComponent<Props, State> {
   private onChange = (params: OnChangeParams): void => {
     if (params.value.length === 0) {
       logWarning("No value selected!")
+      this.setState({ value: undefined })
```
Are you sure that this isn't introducing any issues? In which sitations is this triggered with `params.value.length === 0`? Only when the user clears the value?

sfc-gh-tszerszen replied:
I do not think so, just checked both `st.multiselect` and `st.selectbox` with empty lists/tuples and everything works fine, however we can double check with @jrieke

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -135,20 +141,26 @@ class Selectbox extends React.PureComponent<Props, State> {
       {
         label:
-          options.length > 0
+          options.length > 0 && this.state.value
```
Don't we have to check for null or undefined here? Otherwise this won't show a label when the state is `0`, or?

sfc-gh-tszerszen replied:
Good call out, changed it to `isNullOrUndefined`

lukasmasuch commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -135,20 +141,26 @@ class Selectbox extends React.PureComponent<Props, State> {
+            : clearable
+            ? null
+            : noOptionsMsg,
+        value: this.state.value
```
Same here. Don't we need to check for `None` or `undefined` otherwise this won't work with `0`?

sfc-gh-tszerszen replied:
True, have changed it 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -399,7 +426,7 @@ def _time_input(
         self,
         label: str,
-        value: DateValue = None,
+        value: DateValue | str | None = "...",
```
I think Johannes was referring to the ellipsis object (`...`) and not str `"..."` in the spec (e.g. see some info [here](https://realpython.com/python-ellipsis/)). I don't think we want to add `str` here as an additional allowed type. Using ellipsis this would probably look something like this: `value: DateValue | None  | type(...)  = ...`

Or if we define this on top as a type:

```
DefaultValue = type(...)

# Other code

def date_input(
        self,
        label: str, 
        value: DateValue | DefaultValue | None  = ...
```

sfc-gh-tszerszen replied:
As we discussed we need to use "buriltins.elipsis" here and the PR is updated with that 👍

sfc-gh-tszerszen committed:
Allow setting None as Initial widget value for st.selectbox, st.date_input, st.time_input, st.number_input, st.text_input, st.text_area, st.radio

sfc-gh-mnowotka commented on frontend/src/app/components/StreamlitDialog/SettingsDialog.tsx:
```diff
@@ -178,10 +178,10 @@ export class SettingsDialog extends PureComponent<Props, UserSettings> {
     const { activeTheme: oldTheme, availableThemes }: LibContextProps =
       this.context
-    const newTheme = availableThemes[index]
+    const newTheme = availableThemes[index !== undefined ? index : 0]
```
nit: `const newTheme = availableThemes[index || 0];`

sfc-gh-tszerszen replied:
Updated 👍

sfc-gh-mnowotka commented on frontend/src/lib/components/shared/Dropdown/Selectbox.tsx:
```diff
@@ -20,24 +20,31 @@ import { Select as UISelect, OnChangeParams, Option } from "baseui/select"
   label?: string | null
   labelVisibility?: LabelVisibilityOptions
   help?: string
+  theme: EmotionTheme
```
If `theme` is now required, isn't this a breaking change?

sfc-gh-tszerszen replied:
It will be injected using `import { withTheme } from "@emotion/react"`, later on we have `export default withTheme(Selectbox)`, so it will be always present

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -62,7 +62,10 @@ function stringsToDates(strings: string[]): Date[] {
 /** Convert an array of dates to an array of strings. */
-function datesToStrings(dates: Date[]): string[] {
+function datesToStrings(dates: Date[] | null | undefined): string[] {
+  if (dates === null || dates === undefined) {
```
nit: I quess you can rewrite the condition as `if (dates == null)` or use the `isNullOrUndefined` function you use in a different place in the code.

sfc-gh-tszerszen replied:
`isNullOrUndefined` is perfect here, and I've overlooked it during latest refactor 👍 thank you 👍

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -169,32 +184,55 @@ class NumberInput extends React.PureComponent<Props, State> {
   private commitWidgetValue = (source: Source): void => {
-    const { value } = this.state
+    let value = this.state.value
+    if (this.state.value === undefined || this.state.value === null) {
```
nit: again I think we can either write `if (this.state.value == null)` or use `isNullOrUndefined `.

sfc-gh-tszerszen replied:
True 👍 just went through the code and found a few more places in which I should have used `isNullOrUndefined`, just updated it 👍

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -169,32 +184,55 @@ class NumberInput extends React.PureComponent<Props, State> {
-        value: valueToBeSaved,
-        formattedValue: this.formatValue(valueToBeSaved),
-      })
+      if (value === null || Number.isNaN(value)) {
```
I'm lost in this logic. We already checked `if (value === null || Number.isNaN(value))` few lines above in line 204, correct? Can we maybe check at the beginning of this function if `value` is none or undefined, then handle this situation and then in the rest of the code we know it has some usable value.

sfc-gh-mnowotka replied:
For example check this:

```
private commitWidgetValue = (source: Source): void => {
    const { element, widgetMgr } = this.props;
    const { clearable } = element;
    const data = element;

    let value = this.state.value || null;
    const min = this.getMin();
    const max = this.getMax();
    const node = this.inputRef.current;

    if (value && (min > value || value > max) && node) {
      node.reportValidity();
      return;
    }

    const valueToBeSaved = value || value === 0 ? value : data.default;
    const isValueInvalid = value === null || Number.isNaN(value);
    const isIntData = this.isIntData();

    if (isValueInvalid) {
      widgetMgr.clearIntValue(element, source);
    } else if (isIntData) {
      widgetMgr.setIntValue(element, valueToBeSaved, source);
    } else {
      widgetMgr.setDoubleValue(element, valueToBeSaved, source);
    }

    let newState = {
      dirty: false,
      value: isValueInvalid ? (clearable ? undefined : this.initialValue) : valueToBeSaved,
      formattedValue: this.formatValue(isValueInvalid ? (clearable ? valueToBeSaved : (this.initialValue || 0)) : valueToBeSaved),
    };

    this.setState(newState);
}

```

sfc-gh-tszerszen replied:
Yeah, these are leftovers before refactor to `isNullOrUndefined`, we just need to check `Number.isNaN(value)` nothing more 👍 removed the logic which looked for nulls

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -273,12 +311,14 @@ class NumberInput extends React.PureComponent<Props, State> {
   /** True if the input's current value can be decremented by its step. */
   private get canDecrement(): boolean {
-    return this.state.value - this.getStep() >= this.getMin()
+    const value = this.state.value ? this.state.value : 0
```
what's the `value` type? Is it:

```
const value: int
```
?

If yes, then why do we check something about it on lines 336 and  349?

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -356,15 +396,42 @@ class NumberInput extends React.PureComponent<Props, State> {
             type="number"
             inputRef={this.inputRef}
-            value={formattedValue}
+            value={
```
nit: `value={this.state.value ?? formattedValue}`

sfc-gh-tszerszen replied:
Updated 👍

sfc-gh-mnowotka commented on frontend/src/lib/components/widgets/Selectbox/Selectbox.tsx:
```diff
@@ -83,11 +86,19 @@ class Selectbox extends React.PureComponent<Props, State> {
-      this.state.value,
-      source
-    )
+    if (
```
```
/** Commit state.value to the WidgetStateManager. */
private commitWidgetValue = (source: Source): void => {
  const { element, widgetMgr, clearable } = this.props
  const { value } = this.state
  const data = this.props.element
  const valueToBeSaved = value || value === 0 ? value : data.default

  // Set value to null if it's undefined or null
  value = value ?? null

  const min = this.getMin()
  const max = this.getMax()

  if (value && (min > value || value > max)) {
    const node = this.inputRef.current
    node?.reportValidity()
  } else {
    if (value === null || Number.isNaN(value)) {
      widgetMgr.clearIntValue(element, source)
    } else if (this.isIntData()) {
      widgetMgr.setIntValue(element, valueToBeSaved, source)
    } else {
      widgetMgr.setDoubleValue(element, valueToBeSaved, source)
    }

    const stateUpdates = {
      dirty: false,
      value: (value === null || Number.isNaN(value)) && !clearable ? this.initialValue : valueToBeSaved,
      formattedValue: this.formatValue(valueToBeSaved),
    }

    if ((value === null || Number.isNaN(value)) && clearable) {
      stateUpdates.value = undefined
    }

    this.setState(stateUpdates)
  }
}

```

sfc-gh-mnowotka commented on lib/streamlit/elements/text_widgets.py:
```diff
@@ -253,41 +295,62 @@ def _text_input(
+            try:
+                nullable = (
+                    True
+                    if clearable and ctx.session_state[not_nullable_key] is False
```
nit:
```
nullable = clearable and not ctx.session_state[not_nullable_key]
```

sfc-gh-tszerszen replied:
It's updated 👍

sfc-gh-mnowotka commented on lib/streamlit/elements/number_input.py:
```diff
@@ -15,7 +15,7 @@
 import numbers
```
This whole file requires a serious refactoring.

sfc-gh-tszerszen replied:
Some `if ... else` logic could be shortened just done it 👍 Should we refactor whole file though?

sfc-gh-mnowotka reviewed (commented):
Left some comments to improve the PR.

sfc-gh-mnowotka commented:
In general, I have an impression that we are applying some small changes here and there in the code which increases its complexity and makes it look more spaghetti. But we don't think about a larger theme to refactor it. Maybe this is OK considering the deadlines. However, from a pure engineering perspective and putting the product consideration aside I'd vote to close this PR and rethink the solution.

sfc-gh-tszerszen commented:
> In general, I have an impression that we are applying some small changes here and there in the code which increases its complexity and makes it look more spaghetti. But we don't think about a larger theme to refactor it. Maybe this is OK considering the deadlines. However, from a pure engineering perspective and putting the product consideration aside I'd vote to close this PR and rethink the solution.

Yeah, that's how I implemented it, just added the ability to nullify based on initial value argument, sometimes ex. `text_input` additionally I saved the state somewhere. 

I don't think we cannot avoid `ifs` here, if we stick to the old method and API, and we'd like to keep the widgets behave the same when they are called without `None` as Initial value

We could split underlying code into smaller methods, but that'd require some larger refactor, which I tried to avoid. Also we will affect a lot of widgets

sfc-gh-tszerszen committed:
code shortened and isNullOrUndefined replacement

sfc-gh-tszerszen committed:
fix jstest

sfc-gh-tszerszen committed:
update the app

sfc-gh-tszerszen committed:
change to valueToBeSaved

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -214,17 +228,17 @@ class TimeWidgetsMixin:
         label_visibility: LabelVisibility = "visible",
         step: Union[int, timedelta] = timedelta(minutes=DEFAULT_STEP_MINUTES),
-    ) -> time:
+    ) -> Union[time, None]:
```
nit: `time | None`

sfc-gh-tszerszen replied:
Updated 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -214,17 +228,17 @@ class TimeWidgetsMixin:
     def time_input(
```
We should probably investigate if we can add an overload here to only allow `None` as return if value is  None

sfc-gh-tszerszen replied:
I tried, but didn't really find a way, can look at it in some spare time, however I'd like to put a dot on this, since I'm moving to s4t project

lukasmasuch replied:
> however I'd like to put a dot on this, since I'm moving to s4t project

We probably shouldn't rush this project since it touches many important parts of the codebase :( But it totally understandable that you want to move focus to s4t projects. I can offer to finish this project later this week or next, with you switching to the reviewer role for this PR.

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -512,18 +531,18 @@ def date_input(
         self,
         label: str,
-        value: DateValue = None,
+        value: DateValue | str | None | "builtins.ellipsis" = Ellipsis,
```
Why do we need to add `str` here as type?

sfc-gh-tszerszen replied:
It's a leftover when I used "..." instead of `Ellipsis` as default type, it's updated now 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -399,14 +418,14 @@ def _time_input(
         self,
         label: str,
-        value: DateValue = None,
+        value: DateValue | str | None | "builtins.ellipsis" = Ellipsis,
```
Why do we need to add `str` here as type?

sfc-gh-tszerszen replied:
It's a leftover when I used "..." instead of `Ellipsis` as default type, it's updated now 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -544,12 +563,21 @@ def _date_input(
+        clearable = parsed_values.value is None
+        if clearable:
+            date_input_proto.default[:] = []
+        elif isinstance(parsed_values.value, Iterable) or isinstance(
```
Every `Sequence` is an iterable. So, we can remove this `isinstance(parsed_values.value, Sequence):` check here

sfc-gh-tszerszen replied:
Good to know, it's updated now 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -544,12 +563,21 @@ def _date_input(
-        date_input_proto.min = date.strftime(parsed_values.min, "%Y/%m/%d")
-        date_input_proto.max = date.strftime(parsed_values.max, "%Y/%m/%d")
+        clearable = parsed_values.value is None
+        if clearable:
```
nit: i think its more future proof to use `parsed_values.value is None` here, since we eventually will add a `required` or `clearbable` argument in the future. And the `default` should not be influenced by the parameter, but only by `value` being `None`

sfc-gh-tszerszen replied:
You're right, it's updated 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -214,17 +228,17 @@ class TimeWidgetsMixin:
-        on_change: Optional[WidgetCallback] = None,
-        args: Optional[WidgetArgs] = None,
-        kwargs: Optional[WidgetKwargs] = None,
+        value: TimeValue | None | str | "builtins.ellipsis" = Ellipsis,
```
Why do we need `str` here as type?

sfc-gh-tszerszen replied:
It's a leftover when I used "..." instead of `Ellipsis` as default type, it's updated now 👍

lukasmasuch commented on lib/streamlit/elements/time_widgets.py:
```diff
@@ -353,7 +369,9 @@ def _time_input(
         time_input_proto.label = label
-        time_input_proto.default = time.strftime(parsed_time, "%H:%M")
+        time_input_proto.default = (
+            "" if parsed_time is None else time.strftime(parsed_time, "%H:%M")
```
Empty string here is used to indicate that the default value is empty or? Also probably something to consider to unify via the `optional` proto parameter. In that case, we just don't set `default` if `parsed_time` is `None`.

sfc-gh-tszerszen replied:
Yes, empty string indicates empty value, I tried to avoid changing protobuf field, I am a bit worried about this. In number Input I did not have a choice, since I'd have to pick some number as an indicator, but here `""` empty string seems fine and not invasive.

Are you sure allowing null/undefined here is a good idea?

lukasmasuch commented on lib/streamlit/elements/text_widgets.py:
```diff
@@ -60,9 +82,28 @@ def serialize(self, v: str) -> str:
+    def deserialize(self, ui_value: str | None, widget_id: str = "") -> str:
         return str(ui_value if ui_value is not None else self.value)
 
+    def deserialize_nullable(
```
I'm a bit concerned about this solution here with the flag in `session_state`. The session state is more of a user-facing state; putting some internally used technical flags inside the `session_state` feels problematic. 

Wouldn't it be possible that we make the `value` field optional as well for the text_area / text_input message? And then in the frontend, if `clearable==True`, we always return the widget state with an unset value. It probably will also require changes to the [WidgetStates.proto](https://github.com/streamlit/streamlit/blob/develop/proto/streamlit/proto/WidgetStates.proto) and to the [widget state implementation](https://github.com/streamlit/streamlit/blob/33cd55cdda916d947aa3cfdbc5caee4bb80f2cdf/lib/streamlit/runtime/state/session_state.py#L89) in Python. But I think we can get a cleaner more unified solution implemented here on how widgets can communicate an empty state.

sfc-gh-tszerszen replied:
@LukasMasuch agree its not a clean solution, the problem here is that we need to return `None` initially and `""` empty string afterwards.

The widget keeps the state of only one type and one variable, that how I understand current implementation.

We cannot keep it in proto since on every re-run it will behave the same and on first run we want it to return `None` and later on we want to return `""`. 

It's only the case with text widgets though, other do not require this kind of change, because they do not have this requirement
> if user typed sth before return "" instead of `None`

Do you think changing widget states additionally to all these changes will make this PR less complex? I'm moving to s4t project and I'd like to close it soon

lukasmasuch replied:
> the problem here is that we need to return None initially and "" empty string afterwards.

I think this is something reasonable actually to change in the product spec. I think there is a good argument to make to always return `None` if the text field is empty, in case `clearable` is set to `True`.

sfc-gh-tszerszen replied:
Yeah, removing this from the spec will make implementation much cleaner cc: @jrieke

sfc-gh-tszerszen committed:
typing improvements and minor notices regarding nullable checks

sfc-gh-tszerszen committed:
typing improvements and minor notices regarding nullable checks

lukasmasuch commented:
> However, from a pure engineering perspective and putting the product consideration aside I'd vote to close this PR and rethink the solution.

I think we need a few more iterations here in this PR to figure out how to represent empty state in a unified and consistent way internally across all widgets. Maybe using `optional` for all value fields of the widgets proto messages is a good solution here. But I think it's more of a fine-tuning of this implementation and the room for completely different solutions is a bit limited here.

sfc-gh-tszerszen commented:
> I think we need a few more iterations here in this PR to figure out how to represent empty state in a unified and consistent way internally across all widgets

That would be cool! 

However feels a bit more dangerous than just using already defined proto, ex. "" empty string when date is empty, -1 when Selectbox index is not chosen etc.

Also I think current implementation of widget state stores only one value of one type across reloads

CHerSun commented:
> > I think we need a few more iterations here in this PR to figure out how to represent empty state in a unified and consistent way internally across all widgets
> 
> However feels a bit more dangerous than just using already defined proto, ex. "" empty string when date is empty, -1 when Selectbox index is not chosen etc.
> 
> Also I think current implementation of widget state stores only one value of one type across reloads

Sorry for barging in. I don't know internals there, but thought maybe this could be useful. Have you considered a dataclass approach? They use their own class - _MISSING_TYPE. And 1 constant for it - MISSING (an instance of _MISSING_TYPE, defined as constant in module once). When you define a dataclass decorated class - default=MISSING is used if no default value is supplied to a field.

Personally, I extended that for myself with EMPTY constant (of _MISSING_TYPE too), to make default value explicitly EMPTY (I use several sources of defaults for my code; need to distinguish between just MISSING - need to try to get from another source - and explicit EMPTY = no default, stop the search). Those MISSING / EMPTY values could be translated to more specific value for JS later.

Looks like a good approach to define explicitly missing value (as None could be a legitimate value in some cases; and in the case you described above - there are different values for different cases).
